File: scripts/check-file-size.mjs
#!/usr/bin/env node

/**
 * Check for files exceeding the maximum allowed line count
 * Exits with error code 1 if any files exceed the limit
 *
 * Uses link-foundation libraries:
 * - use-m: Dynamic package loading without package.json dependencies
 */

import { readFileSync, readdirSync, statSync } from 'fs';
import { join, relative, extname } from 'path';

const MAX_LINES = 1000;
const FILE_EXTENSIONS = ['.rs'];
const EXCLUDE_PATTERNS = ['target', '.git', 'node_modules'];

/**
 * Check if a path should be excluded
 * @param {string} path
 * @returns {boolean}
 */
function shouldExclude(path) {
  return EXCLUDE_PATTERNS.some((pattern) => path.includes(pattern));
}

/**
 * Recursively find all Rust files in a directory
 * @param {string} directory
 * @returns {string[]}
 */
function findRustFiles(directory) {
  const files = [];

  function walkDir(dir) {
    const entries = readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(dir, entry.name);

      if (shouldExclude(fullPath)) {
        continue;
      }

      if (entry.isDirectory()) {
        walkDir(fullPath);
      } else if (entry.isFile() && FILE_EXTENSIONS.includes(extname(entry.name))) {
        files.push(fullPath);
      }
    }
  }

  walkDir(directory);
  return files;
}

/**
 * Count lines in a file
 * @param {string} filePath
 * @returns {number}
 */
function countLines(filePath) {
  const content = readFileSync(filePath, 'utf-8');
  return content.split('\n').length;
}

try {
  const cwd = process.cwd();
  console.log(`\nChecking Rust files for maximum ${MAX_LINES} lines...\n`);

  const files = findRustFiles(cwd);
  const violations = [];

  for (const file of files) {
    const lineCount = countLines(file);
    if (lineCount > MAX_LINES) {
      violations.push({
        file: relative(cwd, file),
        lines: lineCount,
      });
    }
  }

  if (violations.length === 0) {
    console.log('All files are within the line limit\n');
    process.exit(0);
  } else {
    console.log('Found files exceeding the line limit:\n');
    for (const violation of violations) {
      console.log(
        `  ${violation.file}: ${violation.lines} lines (exceeds ${MAX_LINES})`
      );
    }
    console.log(`\nPlease refactor these files to be under ${MAX_LINES} lines\n`);
    process.exit(1);
  }
} catch (error) {
  console.error('Error:', error.message);
  process.exit(1);
}

File: scripts/check-file-size.rs
#!/usr/bin/env rust-script
//! Check for files exceeding the maximum allowed line count
//! Exits with error code 1 if any files exceed the limit
//!
//! Usage: rust-script scripts/check-file-size.rs
//!
//! ```cargo
//! [dependencies]
//! walkdir = "2"
//! ```

use std::fs;
use std::path::Path;
use std::process::exit;
use walkdir::WalkDir;

const MAX_LINES: usize = 1000;
const FILE_EXTENSIONS: &[&str] = &[".rs"];
const EXCLUDE_PATTERNS: &[&str] = &["target", ".git", "node_modules"];

fn should_exclude(path: &Path) -> bool {
    let path_str = path.to_string_lossy();
    EXCLUDE_PATTERNS.iter().any(|pattern| path_str.contains(pattern))
}

fn has_valid_extension(path: &Path) -> bool {
    if let Some(ext) = path.extension() {
        let ext_with_dot = format!(".{}", ext.to_string_lossy());
        FILE_EXTENSIONS.contains(&ext_with_dot.as_str())
    } else {
        false
    }
}

fn count_lines(path: &Path) -> Result<usize, std::io::Error> {
    let content = fs::read_to_string(path)?;
    Ok(content.lines().count())
}

struct Violation {
    file: String,
    lines: usize,
}

fn main() {
    println!("\nChecking Rust files for maximum {} lines...\n", MAX_LINES);

    let cwd = std::env::current_dir().expect("Failed to get current directory");
    let mut violations: Vec<Violation> = Vec::new();

    for entry in WalkDir::new(&cwd)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().is_file())
    {
        let path = entry.path();

        if should_exclude(path) {
            continue;
        }

        if !has_valid_extension(path) {
            continue;
        }

        match count_lines(path) {
            Ok(line_count) => {
                if line_count > MAX_LINES {
                    let relative_path = path
                        .strip_prefix(&cwd)
                        .unwrap_or(path)
                        .to_string_lossy()
                        .to_string();
                    violations.push(Violation {
                        file: relative_path,
                        lines: line_count,
                    });
                }
            }
            Err(e) => {
                eprintln!("Warning: Could not read {}: {}", path.display(), e);
            }
        }
    }

    if violations.is_empty() {
        println!("All files are within the line limit\n");
        exit(0);
    } else {
        println!("Found files exceeding the line limit:\n");
        for violation in &violations {
            println!(
                "  {}: {} lines (exceeds {})",
                violation.file, violation.lines, MAX_LINES
            );
        }
        println!("\nPlease refactor these files to be under {} lines\n", MAX_LINES);
        exit(1);
    }
}
